### PR TITLE
slither-read-storage: Handle unstructured storage

### PR DESCRIPTION
**Pull request by webthethird**
This approach scans for constant bytes32 variables and uses a couple heuristics to determine (a) that it is a storage slot, and (b) the type of value stored in the slot.

For example, running the following command yields the table below:
```
slither-read-storage 0x8315177aB297bA92A06054cE80a67Ed4DBd7ed3a --contract TransparentUpgradeableProxy --value --table --rpc-url <MY_RPC_URL>
```
```bash
+----------------------+-------------+-------------------------------------------------------------------------------+------+--------+--------------------------------------------+
|         name         | type_string |                                      slot                                     | size | offset |                   value                    |
+----------------------+-------------+-------------------------------------------------------------------------------+------+--------+--------------------------------------------+
|     _BEACON_SLOT     |   address   | 74152234768234802001998023604048924213078445070507226371336425913862612794704 | 160  |   0    | 0x0000000000000000000000000000000000000000 |
|     _ADMIN_SLOT      |   address   | 81955473079516046949633743016697847541294818689821282749996681496272635257091 | 160  |   0    | 0x554723262467F125Ac9e1cDFa9Ce15cc53822dbD |
| _IMPLEMENTATION_SLOT |   address   | 24440054405305269366569402256811496959409073762505157381672968839269610695612 | 160  |   0    | 0x1066CEcC8880948FE55e427E94F1FF221d626591 |
|    _ROLLBACK_SLOT    |     bool    | 33048860383849004559742813297059419343339852917517107368639918720169455489347 |  1   |   0    |                   False                    |
+----------------------+-------------+-------------------------------------------------------------------------------+------+--------+--------------------------------------------+
```

### CODE REVIEW

**Comment by montyly** [comment]
Can we add a test with an address that uses this format?

**Comment by webthethird** [comment]
> Can we add a test with an address that uses this format?

The address I used is 0x8315177aB297bA92A06054cE80a67Ed4DBd7ed3a, which is the Arbitrum Bridge's Transparent Upgradeable Proxy.

Where would you recommend adding the test? `/scripts/ci_test_upgradeability.sh` currently just tests `slither-check-upgradeability`, and this isn't really part of that tool. But at a glance I don't see a test script for `slither-read-storage`, though I do see the `/tests/storage-layout` directory.

**Comment by 0xalpharush** [comment]
@webthethird You would deploy a contract to ganache and run slither-read-storage against the local network as is done here https://github.com/crytic/slither/blob/master/tests/test_read_storage.py

**Comment by webthethird** [comment]
@0xalpharush would it make sense to add it to the current test? Since this adds to slither-read-storage itself, adding to what it will print if the contract uses any constant storage slots.

**Comment by webthethird** [comment]
@0xalpharush Would it make sense to add it to the existing test? Since this PR applies to slither-read-storage itself, just adding to what would print if the contract used any constant storage slots.

**Comment by webthethird** [comment]
@montyly @0xalpharush I added a new test, `tests/test_unstructured_storage.py`.

I believe the failing CI test is related to #1710

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:279]: You may be able to use https://crytic.github.io/slither/slither/core/declarations/contract.html#Contract.get_functions_reading_from_variable to make this more general and not have to rely on matching these specific function names. If it only returns one function and that function has one return value that's less than or equal to 32 bytes, assume the return type is that.

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:131]: Can you not get the literal expression without this?

File: tests/test_unstructured_storage.py
Comment (0xalpharush) [inline tests/test_unstructured_storage.py:40]: You should be able to import these fixtures without reimplementing them

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:333]: Can you loop over `contract.state_variables_ordered`  once instead of twice here and append the relevant variables to `_target_variables ` and `_constant_storage_slots `?

File: tests/test_unstructured_storage.py
Comment (webthethird) [reply tests/test_unstructured_storage.py:40]: I am trying to do this, but when I import them pylint complains that they are unused, which is incorrect since pytest needs them.
```
tests/test_unstructured_storage.py:9:0: W0611: Unused GanacheInstance imported from tests.test_read_storage (unused-import)
tests/test_unstructured_storage.py:9:0: W0611: Unused fixture_ganache imported from tests.test_read_storage (unused-import)
tests/test_unstructured_storage.py:9:0: W0611: Unused fixture_web3 imported from tests.test_read_storage (unused-import)
```

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:343]: Do you mean like this?
```python
    def get_all_storage_variables(self, func: Callable = None) -> None:
        """Fetches all storage variables from a list of contracts.
        kwargs:
            func (Callable, optional): A criteria to filter functions e.g. name.
        """
        for contract in self.contracts:
            for var in contract.state_variables_ordered:
                if func is None or (func is not None and func(var)):
                    if not var.is_constant and not var.is_immutable:
                        self._target_variables.append((contract, var))
                    elif var.is_constant and str(var.type) == "bytes32":
                        self._constant_storage_slots.append((contract, var))
            hardcoded_slot = self.find_hardcoded_slot_in_fallback(contract)
            if hardcoded_slot is not None:
                self._constant_storage_slots.append((contract, hardcoded_slot))
```

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:131]: Which line(s) are you referring to?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [reply slither/tools/read_storage/read_storage.py:343]: yeah. also `str(var.type) == "bytes32"` can be `var.type == ElementaryType("bytes32")`

File: tests/test_unstructured_storage.py
Comment (0xalpharush) [reply tests/test_unstructured_storage.py:40]: The fixtures need to be moved to a `conftest.py`

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:142]: > Which line(s) are you referring to?
These I believe you can get the initial expression as a literal

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:139]: > Which line(s) are you referring to?
I think you can get the literal value from the expression here

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py]: Alright, I've made these changes

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:139]: Got it, made the change

File: tests/test_unstructured_storage.py
Comment (webthethird) [reply tests/test_unstructured_storage.py]: Would that be a new file specific to `test_unstructured_storage`? Or would we be moving the fixtures from `test_read_storage` and importing `conftest.py` into both tests?

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:281]: >If it only returns one function and that function has one return value that's less than or equal to 32 bytes, assume the return type is that.

For the most part I am doing this on lines 275-278. Though I could streamline it by using `Contract.get_functions_reading_from_variable`.

> to make this more general and not have to rely on matching these specific function names.

The reason I resorted to checking for the function names from the StorageSlot library is because there may be some slot (i.e., `_ROLLBACK_SLOT` in the standard ERC1967Upgrade.sol) which do not have any setter or getter functions. With the `_ROLLBACK_SLOT` example, the only function which reads from that variable is `_upgradeToAndCallUUPS` which contains the line `if (StorageSlot.getBooleanSlot(_ROLLBACK_SLOT).value)`.

File: tests/test_unstructured_storage.py
Comment (webthethird) [reply tests/test_unstructured_storage.py]: I read up a little bit on modular pytest fixtures, and it seems my second interpretation is accurate. Would we also need to move all other fixtures from other test programs? Or would it suffice to use `conftest.py` only for `test_read_storage` and `test_unstructured_storage` for now?

Comment (webthethird) [reply tests/test_unstructured_storage.py]: Assuming the latter is good, this is done.

**Comment by webthethird** [comment]
@0xalpharush Using `conftest.py` seems to be screwing up other CI tests! :grimacing:

**Comment by webthethird** [comment]
@montyly @0xalpharush Are there any remaining blockers on this PR? I'd like to start using it in combination with #1757

**Comment by montyly** [comment]
@0xalpharush: can you do another pass on this PR?

File: tests/storage-layout/StorageSlot.abi
Comment (0xalpharush) [inline tests/storage-layout/StorageSlot.abi:1]: Can you remove this?

File: tests/conftest.py
Comment (0xalpharush) [inline tests/conftest.py:1]: Can you just put all the fixtures here instead of importing them?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:266]:
```suggestion
        for c in self.contracts_derived:
```

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:297]: I'm not really sure how this will hold up to any contract on mainnet. I'd almost rather leave the type as bytes32... What is this used for?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:355]: I don't think this works if the argument is not a literal

File: tests/fixtures/fixture_web3.py
Comment (0xalpharush) [inline tests/fixtures/fixture_web3.py:6]: This try-except can be removed now that web3 is a dependency

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:276]: `ret` here is a Type already, and not all Type subclasses have a `.type` property it seems. I just return `str(ret)` since all that's needed is a type string. But I am open to suggestions.

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:355]: Yes, but finding that hardcoded Literal is the purpose of this function. I can't think of another circumstance where a constant slot isn't covered by either the for loop in `get_all_storage_variables` or the subsequent call to `find_hardcoded_slot_in_fallback`. If you think of one though I'd like to make sure it's covered!

File: tests/storage-layout/StorageSlot.abi
Comment (webthethird) [reply tests/storage-layout/StorageSlot.abi:1]: And put it in conftest.py instead, right?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [reply slither/tools/read_storage/read_storage.py:355]: I mean that the parsing does not handle the case where it is something like `sload(add(1,1))`

Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:355]: Good point. Granted, this part of the code (with the string parsing) is only meant for solc <0.6.0, where the assembly string is all we've got. Handling all of the corner cases with string parsing might get pretty ugly. However, the portion starting at line 364 with `elif node.type == NodeType.EPRESSION:` is meant to handle more recent solc versions, so this should definitely work if it's not a Literal. Let me test it a bit more.

Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:355]: Given that this is already handling a bit of a corner case (usually storage slots are declared as constants), how robust does this need to be?

Comment (0xalpharush) [reply slither/tools/read_storage/read_storage.py:355]: The concern is that this breaks regular usage of slither-read-storage, so it should not fail if it comes across code it cannot parse

Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:355]: Would it be sufficient for `find_hardcoded_slot_in_fallback` to return nothing if it finds code it can't parse?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:149]: When is this `TypeError` being hit? It seems like it could be removed by performing more validations of what type `exp.value` is

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:261]: Can you use ElementaryType("bytes32")

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:367]: This can be a nested expression. I wonder if this code could benefit from reusing `ConstantFolding` to account for nested expressions

**Comment by 0xalpharush** [comment]
My main concern is that the code handles the happy path. We need to add tests for https://github.com/crytic/slither/pull/1752/commits/40bf399a46667a0bdca58b7131653d1248969bf6 and cases like https://github.com/crytic/slither/pull/1752#discussion_r1146349516

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:261]: Oh right, I guess I didn't think to check if `Type.__eq__()` was implemented

**Comment by montyly** [comment]
@0xalpharush: let me know when this is ready.

@webthethird: for future PRs, I would recommend to split the changes into individual PRs. For example here the changes to the constant folding could have been its own PR, and the change on the read-storage a second PR (build on top of the first PR branch). Having small PRs makes the review easier/faster ;)

**Comment by webthethird** [comment]
> for future PRs, I would recommend to split the changes into individual PRs. For example here the changes to the constant folding could have been its own PR, and the change on the read-storage a second PR (build on top of the first PR branch).

Got it, I'll be more cognizant of that going forward! For what it's worth, the change to constant folding was its own PR, created by @0xalpharush after I made this one, which enabled handling one of the unhappy paths when finding constant storage slots. So I merged that into this PR, though I probably should have started a second PR built on top of this one, and done the merge there.

**Comment by 0xalpharush** [comment]
@webthethird It doesn't look like you merge correctly updated `setup.py`

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:146]: The `NotConstant` error is not being handling here

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:280]: If this is a precondition, should it instead be an assert?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:331]: Was this not something that could be replaced by the constant folding code?

**Comment by 0xalpharush** [comment]
Can we put this behind a flag that way its opt-in and will be skipped if its not necessary?

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:141]: I didn't think ConstantFolding would accept a CallExpression, since we have
```python
CONSTANT_TYPES_OPERATIONS = Union[
    Literal, BinaryOperation, UnaryOperation, Identifier, TupleExpression, TypeConversion
]
```
So these two lines extract the inner expression, which can then fall through to the following lines where ConstantFolding is used.

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:146]: If the expression is not constant then I suppose the variable in question can't be a constant storage slot, so we can skip it by adding `NotConstant` to the except clause on line 171:
```python
    def get_unstructured_layout(self) -> Dict[str, SlotInfo]:
        tmp: Dict[str, SlotInfo] = {}
        for _, var in self.constant_slots:
            var_name = var.name
            try:
                exp = var.expression
                if isinstance(exp, CallExpression) and str(exp.called) == "keccak256(bytes)":
                    exp = exp.arguments[0]
                if isinstance(
                    exp,
                    (BinaryOperation, UnaryOperation, Identifier, TupleExpression, TypeConversion),
                ):
                    exp = ConstantFolding(exp, "bytes32").result()
                ...
            except (TypeError, NotConstant):
                continue
```

**Comment by webthethird** [comment]
> Can we put this behind a flag that way its opt-in and will be skipped if its not necessary?

Yes, absolutely

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:292]: What is this for?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:303]:
```suggestion
        for func in funcs:
            if func.return_type:
              for ret in rets:
                  size, _ = ret.storage_size
                  if size <= 32:
                      return str(ret), size * 8
```
You probably can't accept this directly due to the whitespace btw

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:377]: Should this also be guarded by the unstructured flag?

File: tests/tools/read-storage/test_data/UnstructuredStorageLayout.abi
Comment (0xalpharush) [inline tests/tools/read-storage/test_data/UnstructuredStorageLayout.abi:1]: I don't think the abi is needed.

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:480]: Is this needed or can be removed?

File: slither/tools/read_storage/read_storage.py
Comment (0xalpharush) [inline slither/tools/read_storage/read_storage.py:489]: Can you not use `decode` to convert bytes to`str`?

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:292]: This is for when a function reads from a storage slot without using the defined constant var, but using the same expression, i.e., 
```solidity
contract UnstructuredStorageLayout {
    bytes32 constant BEACON_SLOT = bytes32(uint256(keccak256('eip1967.proxy.beacon')) - 1);
    ...
    function _set_beacon(address _beacon) internal {
        bytes32 slot = bytes32(uint256(keccak256('eip1967.proxy.beacon')) - 1);
        assembly {
            sstore(slot, _beacon)
        }
    }
    ...
}
```

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:303]: I believe I wrote this the way I did in order to avoid a pylint `too-many-branches` complaint, since I'm right there at the maximum of 12 branches. Adding `if func.return_type` puts it at 13.

Should I leave it this way, or disable `too-many-branches` for this function?

File: tests/tools/read-storage/test_data/UnstructuredStorageLayout.abi
Comment (webthethird) [reply tests/tools/read-storage/test_data/UnstructuredStorageLayout.abi:1]: In my test code, it seems like `test_read_storage.deploy_contract` uses the abi:
```python
@pytest.mark.usefixtures("web3", "ganache")
def test_unstructured_storage(web3, ganache) -> None:
    assert web3.is_connected()
    bin_path = Path(TEST_DATA_DIR, "UnstructuredStorageLayout.bin").as_posix()
    abi_path = Path(TEST_DATA_DIR, "UnstructuredStorageLayout.abi").as_posix()
    bytecode = get_source_file(bin_path)
    abi = get_source_file(abi_path)
    contract = deploy_contract(web3, ganache, bytecode, abi)
    ...
```

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:480]: I think it's needed. This is basically the definition of the hardcoded slot that `find_hardcoded_slot_in_exp` is looking for.

I know ConstantFolding always returns a Literal (if it doesn't raise a NotConstant exception), but there's no guarantee that ConstantFolding is used in the lines above this, hence the check for `isinstance(exp, Literal)`. Perhaps I could get rid of the `isinstance(exp.type, ElementaryType)`, since it appears that will always be the case when ConstantFolding is used. But then I think I need the `exp.type.name in ["bytes32", "uint256"]` because I'm not interested in constant addresses, strings or bools.

Does that make sense?

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:497]: Is it possible for a variable in an abstract contract to end up in `self.target_variables`?

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:489]: It seems `bytes.hex()` works best here

File: slither/tools/read_storage/read_storage.py
Comment (webthethird) [reply slither/tools/read_storage/read_storage.py:377]: It could be. I just figured it didn't hurt to add the variables to `self._constant_storage_slots` anyway, since this part is not too computationally complex.

**Comment by webthethird** [comment]
@0xalpharush After merging from dev I'm getting some odd test errors. Any idea why?

Also please let me know if there's anything else to do on this PR to get it ready to merge. This one has probably been around too long!

File: tests/tools/read-storage/test_read_storage.py
Comment (0xalpharush) [inline tests/tools/read-storage/test_read_storage.py:102]: This is missing `solc_binary_path` which handles installing the solc dependency. The code should look something like https://github.com/crytic/slither/blob/4149f6034163e6252a6a277ccd03d811701ec757/tests/tools/read-storage/test_read_storage.py#L93-L105

**Comment by webthethird** [comment]
Now I'm getting an error when trying to run Black
```
Error: Unable to resolve action `github/super-linter@v4.9.2`, unable to find version `v4.9.2`
```

**Comment by webthethird** [comment]
With the Black issue resolved, all checks are passing

**Review by 0xalpharush — APPROVED** [review]

**Comment by 0xalpharush** [comment]
I think we can squash and merge

**Comment by 0xalpharush** [comment]
@webthethird Can you resolve merge conflicts?

**Comment by webthethird** [comment]
@0xalpharush merge conflicts resolved, and tests passing

File: slither/visitors/expression/constants_folding.py
Comment (montyly) [inline slither/visitors/expression/constants_folding.py:67]: mhh would this cause problem for `bytes1`/ `bytes2`/ ...

Comment (webthethird) [reply slither/visitors/expression/constants_folding.py:67]: @0xalpharush what do you think? This is part of the constant folding change you made I think.

**Comment by webthethird** [comment]
@0xalpharush I just merged dev into this branch again, and suddenly we're getting a bunch of test failures. Any idea what the cause might be? I had already merged from dev yesterday, so it must be something recent.

**Pull request closed without merging**
